frontend/src/home/graph-d3.js

```diff
@@ -0,0 +1,427 @@
+var d3 = require("d3");
+var slider = require("d3-simple-slider");
+
+
+// Helper functions
+function get_group_map(links) {
+    // create a map from node_id to node
+    let group_map = {};
+    links.forEach(d => {
+        group_map[d.source] = null;
+        group_map[d.target] = null;
+    });
+
+    // create a map from node.id to directly connected node.ids
+    let link_map = {};
+    links.forEach(d => {
+        if (d.source in link_map) {
+            link_map[d.source].push(d.target);
+        } else {
+            link_map[d.source] = [d.target];
+        }
+
+        if (d.target in link_map) {
+            link_map[d.target].push(d.source);
+        } else {
+            link_map[d.target] = [d.source];
+        }
+    })
+
+    let group_id = 0;
+    let visited = new Set();
+    let unseen = new Set(Object.keys(group_map));
+
+    // visit all nodes
+    while (unseen.size > 0) {
+        // start with an unseen node
+        let node = unseen.values().next().value;
+        unseen.delete(node);
+
+        // DFS for all reachable nodes, start with all directly reachable
+        let stack = [node];
+        while (stack.length > 0) {
+            // take top node on stack
+            let cur_node = stack.pop();
+
+            // visit it, give it a group
+            visited.add(cur_node);
+            group_map[cur_node] = group_id;
+
+            // add all directly reachable (and unvisited nodes) to stack
+            let directly_reachable = link_map[cur_node].filter(n => unseen.has(n));
+            stack = stack.concat(directly_reachable);
+
+            // remove all directly reachable elements from unseen
+            directly_reachable.forEach(n => unseen.delete(n));
+        }
+
+        group_id++;
+    }
+
+    return group_map;
+}
+
+function get_real_width(elem, props) {
+    // if we're given a static width, use it
+    if (props.width !== undefined) {
+        return props.width;
+    }
+
+    // calculate width of provided (likely, parent) element
+    let style = getComputedStyle(elem);
+    return elem.clientWidth - parseFloat(style.paddingLeft) - parseFloat(style.paddingRight);
+}
+
+
+// Graph -- store in an object to maintain state
+var d3Graph = {
```

Comment:
`d3Graph` should probably be a ` class` instead of an object at module level scope. That should enable us to use multiple graphs in the same page. (I don't know why we ever would but hey)